### PR TITLE
kernel: Modify the signature of `k_mem_slab_free` via deprecation

### LINKED ISSUE #61888: `k_mem_slab_free()` should take `void *mem` instead of `void **mem`
**Describe the bug**

The following function:
```c
void k_mem_slab_free(struct k_mem_slab *slab, void **mem);
```
has the wrong signature. `mem` is only used as a regular pointer, so there is no need to use a double-pointer. The correct signature should be:

```c
void k_mem_slab_free(struct k_mem_slab *slab, void *mem);
```

The issue with the current signature, although functional, is that it is extremely confusing. I myself, a veteran Zephyr developer, was confused by this parameter when looking at it recently.

**Impact**

This impacts the developer in two ways:

1. It requires the developer to unnecessarily compute the address with `&` when calling the function
2. It confuses the user by implying that the pointer itself will be modified

**Additional context**

Note that `k_mem_slab_alloc()` takes a double pointer, but in that case it makes sense, since the result is stored in it.

### PR DESCRIPTION
**Pull request by carlescufi**
Deprecate the current `k_mem_slab_free()` function to be able to introduce
a new `k_mem_slab_free2()` version with the correct signature, replacing
the old void `**mem` with void `*mem` as a parameter.

Fixes #61888.

Please see https://github.com/zephyrproject-rtos/zephyr/pull/61901 for an alternative that uses the stable API change procedure

### CODE REVIEW

**Comment by gmarull** [comment]
Sad to see this “new” API. Stability cult has gone too far, sorry but can’t approve this one.

File: doc/kernel/memory_management/slabs.rst [hunk 1]
```diff
@@ -119,7 +119,7 @@ A warning is printed if a suitable block is not obtained.
 Releasing a Memory Block
 ========================
 
-A memory block is released by calling :c:func:`k_mem_slab_free`.
+A memory block is released by calling :c:func:`k_mem_slab_free2`.
```
[the remark this answers is not in the document]
Comment (henrikbrixandersen) [reply doc/kernel/memory_management/slabs.rst:122]: I agree. Versioned APIs do not scale.

**Review by andyross — DISMISSED** [review]
[I swear I just typed up a review and submitted but now I can't find it, github is haunted sometimes]

This looks fine to me, but I will say I prefer the other two options you mentioned.  Putting a "2" in the function name means that apps doing The Right Thing and writing to the official API will be forced to change again when we come back post-deprecation and reclaim the bare "k_mem_slab_free()" symbol.

I guess my ideal preference would be a "COMPAT_MEM_SLAB" kconfig or whatever, defaulting to n, which apps that want to use the deprecated value need to set.  Then we can get new code on the new API faster.

But no complaints in particular with this method either.   Better that we do the API hygiene work than dither over it.

File: doc/kernel/memory_management/slabs.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (fabiobaltieri) [reply doc/kernel/memory_management/slabs.rst:122]: 😱

**Comment by carlescufi** [comment]
> This looks fine to me, but I will say I prefer the other two options you mentioned.

So do I, and others. This PR is here to open the discussion, great to have your feedback @andyross!

File: doc/kernel/memory_management/slabs.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (carlescufi) [reply doc/kernel/memory_management/slabs.rst:122]: That is what we are here to decide, opinions welcome!

[the remark this answers is not in the document]
Comment (carlescufi) [reply doc/kernel/memory_management/slabs.rst:122]: I don't particularly like them either, but they are widely used (see for example `mmap` and `mmap2` in the kernel).

[the remark this answers is not in the document]
Comment (jfischer-no) [reply doc/kernel/memory_management/slabs.rst:122]: :scream: :scream: :scream: :scream: :scream: :scream: :scream: :scream: :scream: :scream: :scream: :scream:

[the remark this answers is not in the document]
Comment (gmarull) [reply doc/kernel/memory_management/slabs.rst:122]: image

**Comment by carlescufi** [comment]
Superseeded by https://github.com/zephyrproject-rtos/zephyr/pull/61901

**Pull request closed without merging**